### PR TITLE
🚀 Webapp docker service and testing using composite actions

### PR DESCRIPTION
**Pull request by rahulpatidar0191**
This action runs docker up and then perform tests on them and then removes them once it's finished testing

Scripts are in [one folder](https://github.com/SatelCreative/satel-python-docker-tests/tree/feature/webapp-deployment-shell/scripts)

Closes: https://github.com/SatelCreative/DevOps-IT/issues/205

**Tested on** 
1. [ST-PIM](https://github.com/SatelCreative/st-pim/blob/ee35fbe167aba81aaa42faec2a6e9cfc3aba2668/.github/workflows/test-and-deploy-webapp.yml#L827)
2. [SB-PIM](https://github.com/SatelCreative/sb-pim/blob/669f8fd6d0189dc0f3ad9672974352cc7c1ea0b1/.github/workflows/test-and-deploy-webapp.yml#L81)

- The [action.yml](https://github.com/SatelCreative/satel-webapp-testing/blob/feature/webapp-deployment-shell/action.yml) creates an action
- Readme shows the usage of the actions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,37 @@
...
+  push:
+    tags:
+      - "*"
+    branches:
+      - main  
+    
+    build-server:
+    name: Build server
+    needs: [generate-variables, build-client]
+    runs-on: <host_name>
```
Comment (hillairet) [inline README.md]: Make it more visible.
Also how about putting the docs at the same place?
```suggestion
    # HOST_NAME is set to "self-hosted" or the name of server where the action runner is hosted, "cosmicray" for example
    runs-on: <HOST_NAME>
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,37 @@
...
+    steps:
+     - name: Local docker environment
+        uses: SatelCreative/satel-local-docker-env@v1
+        with:
+          app-name: <app-name>
```
Comment (hillairet) [inline README.md]: I would keep consistency on the use of `_` or `-` in the names of variables.
```suggestion
          app-name: <APP_NAME> 
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,37 @@
-# satel-webapp-testing
-Centralized GitHub actions to test a webapp
+# Satel Local Docker environment
+This centralized GitHub action runs tests on local docker services
+
+## Usage 
+```yml
+name: "Deploy webapp"
```
Comment (hillairet) [inline README.md]: Just a name but doesn't match the rest
```suggestion
name: "build and deploy webapp"
```

File: scripts/CodeTestsMultiple.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+docker-compose exec -T pim_api python -c "import requests; requests.get('http://localhost:8000/health')" || exit 1
+
+echo "Clean up old reports" 
+rm -f unittesting.xml coverage.xml typing-server.xml typing-integrations.xml
+
+echo "Code tests" 
+## Catch the exit codes so we don't exit the whole script before we are done.
+## Typing, linting, formatting check & unit and integration testing
+docker-compose ps
+docker-compose exec -T pim_api flake8; STATUS1=$?
```
Comment (hillairet) [inline scripts/CodeTestsMultiple.sh]: Wait why don't use use `validatecodeonce` here and do only `flake8`?

Comment (rahulpatidar0191) [reply scripts/CodeTestsMultiple.sh]: This was a special case for [sb-pim](https://github.com/SatelCreative/sb-pim/blob/6cf9b95c5fd82796d19b0bd7cc961d4de8ed3275/Jenkinsfile#L108), I just took what was done in that repo

File: README.md (hunk 2 above)
Comment (rahulpatidar0191) [reply README.md]: Yep! I need to do this at few places. Definitely want to be consistent before it gets messy

File: scripts/CodeTestsMultiple.sh (hunk 4 above)
Comment (rahulpatidar0191) [reply scripts/CodeTestsMultiple.sh]: @hillairet, I ran the `validatecodeonce` on sb-pim and there were [many errors](https://github.com/SatelCreative/sb-pim/actions/runs/4600104241/jobs/8126727283) with it. Let me know if I should contact frank to fix them or just keep it running with `flake8`

Comment (hillairet) [reply scripts/CodeTestsMultiple.sh]: That's the lack of flexibility with this scheme that is not ideal ... flake8 only is fine. But Frank should fix the rest at some point.

Comment (rahulpatidar0191) [reply scripts/CodeTestsMultiple.sh]: I have reported this to SB-PIM management team and ask them to put in the backlog as they don't have budget for it atm. Frank said it'll be 2 days of work to fix them

File: scripts/CodeTestsOnce.sh [hunk 5]
```diff
@@ -8,23 +8,23 @@ fi
...
 ## Catch the exit codes so we don't exit the whole script before we are done.
 ## Typing, linting, formatting check & unit and integration testing
-docker-compose exec -T webapp validatecodeonce; STATUS1=$?
-docker cp "$(docker-compose ps -q webapp)":/python/reports/typing.xml typing.xml
-docker cp "$(docker-compose ps -q webapp)":/python/reports/unittesting.xml unittesting.xml
-docker cp "$(docker-compose ps -q webapp)":/python/reports/coverage.xml coverage.xml
+docker-compose exec -T ${CONTAINER_NAME} validatecodeonce; STATUS1=$?
+docker cp "$(docker-compose ps -q ${CONTAINER_NAME})":/python/reports/typing.xml typing.xml
+docker cp "$(docker-compose ps -q ${CONTAINER_NAME})":/python/reports/unittesting.xml unittesting.xml
+docker cp "$(docker-compose ps -q ${CONTAINER_NAME})":/python/reports/coverage.xml coverage.xml
```
Comment (hillairet) [inline scripts/CodeTestsOnce.sh]: Are these files used by Github? They were useful in Jenkins ...

Comment (rahulpatidar0191) [reply scripts/CodeTestsOnce.sh]: Not yet, but I'm hoping to use them with something like this https://github.com/marketplace/actions/junit-report-action

**Comment by rahulpatidar0191** [comment]
@hillairet Last one

**Review by hillairet — APPROVED** [review]
Let's do this

**Pull request merged**